src/messages/mod.rs

```diff
@@ -15,29 +18,100 @@ mod test_proposals;
 #[cfg(test)]
 mod test_welcome;
 
+/// Welcome Messages
+///
+/// > 11.2.2. Welcoming New Members
+///
+/// ```text
+/// struct {
+///   ProtocolVersion version = mls10;
+///   CipherSuite cipher_suite;
+///   EncryptedGroupSecrets secrets<0..2^32-1>;
+///   opaque encrypted_group_info<1..2^32-1>;
+/// } Welcome;
+/// ```
+#[derive(Clone, Debug, PartialEq)]
+pub struct Welcome {
+    version: ProtocolVersion,
+    cipher_suite: &'static Ciphersuite,
+    secrets: Vec<EncryptedGroupSecrets>,
+    encrypted_group_info: Vec<u8>,
+}
+
+/// EncryptedGroupSecrets
+///
+/// > 11.2.2. Welcoming New Members
+///
+/// ```text
+/// struct {
+///   opaque key_package_hash<1..255>;
+///   HPKECiphertext encrypted_group_secrets;
+/// } EncryptedGroupSecrets;
+/// ```
+#[derive(Clone, Debug, PartialEq)]
+pub struct EncryptedGroupSecrets {
+    pub key_package_hash: Vec<u8>,
+    pub encrypted_group_secrets: HpkeCiphertext,
+}
+
+impl Welcome {
+    /// Create a new welcome message from the provided data.
+    /// Note that secrets and the encrypted group info are consumed.
+    pub(crate) fn new(
+        version: ProtocolVersion,
+        cipher_suite: &'static Ciphersuite,
+        secrets: Vec<EncryptedGroupSecrets>,
+        encrypted_group_info: Vec<u8>,
+    ) -> Self {
+        Self {
+            version,
+            cipher_suite,
+            secrets,
+            encrypted_group_info,
+        }
+    }
+
+    /// Get a reference to the ciphersuite in this Welcome message.
+    pub(crate) fn ciphersuite(&self) -> &'static Ciphersuite {
+        self.cipher_suite
+    }
+
+    /// Get a reference to the ciphersuite in this Welcome message.
+    pub fn get_secrets_ref(&self) -> &[EncryptedGroupSecrets] {
```

Comment:
```suggestion
    pub fn secrets(&self) -> &[EncryptedGroupSecrets] {
```
or maybe `secrets_ref`, because it returns a slice?

Reply:
Yeah, this is all just c&p I didn't actually touch the code. But I can do these changes (as we want them).